File: LeetCodeProblemsCpp/LeetCodeProblemsCpp/Problems/LeetCode1717MaximumScoreFromRemovingSubstrings.cpp
#include "LeetCode1717MaximumScoreFromRemovingSubstrings.h"

#include <stack>
using namespace std;
int LeetCode1717MaximumScoreFromRemovingSubstrings::maximumGain( string s, int x, int y)
{
    stack<char> remaining{};
    int score{};
    string firstTarget=x>y?"ab":"ba";
    int firstScore=x>y?x:y;
    string secondTarget=x>y?"ba":"ab";
    int secondScore=x>y?y:x;
    for (char c:s)
    {
        if (remaining.empty())
        {
            remaining.push(c);
            continue;
        }
        auto top=remaining.top();
        if (top==firstTarget[0] && c==firstTarget[1])
        {
            remaining.pop();
            score+=firstScore;
            continue;
        }
        remaining.push(c);
    }
    stack<char> secondRemaining{};
    while (!remaining.empty())
    {
        auto c=remaining.top();
        remaining.pop();
        if (secondRemaining.empty())
        {
            secondRemaining.push(c);
            continue;
        }
        auto top=secondRemaining.top();
        if (top==secondTarget[1] && c==secondTarget[0])
        {
            secondRemaining.pop();
            score+=secondScore;
            continue;
        }
        secondRemaining.push(c);
    }
    return score;
}

File: LeetCodeProblemsGo/LeetCode1717MaximumScoreFromRemovingSubstrings/LeetCode1717MaximumScoreFromRemovingSubstrings.go
package LeetCode1717MaximumScoreFromRemovingSubstrings

func maximumGain(s string, x int, y int) int {
	numsStack := &myStack{}
	score := 0
	firstScore := max(x, y)
	firstTarget := "ab"
	secondTarget := "ba"
	if y > x {
		secondTarget = "ab"
		firstTarget = "ba"
	}
	secondScore := min(x, y)
	for _, c := range s {
		if numsStack.isEmpty() {
			numsStack.push(c)
			continue
		}
		if numsStack.peek() == rune(firstTarget[0]) && c == rune(firstTarget[1]) {
			score += firstScore
			numsStack.pop()
			continue
		}
		numsStack.push(c)
	}
	secondStack := &myStack{}
	for !numsStack.isEmpty() {
		c := numsStack.pop()
		if secondStack.isEmpty() {
			secondStack.push(c)
			continue
		}
		if secondStack.peek() == rune(secondTarget[1]) && c == rune(secondTarget[0]) {
			score += secondScore
			secondStack.pop()
			continue
		}
		secondStack.push(c)
	}
	return score
}

type myStack struct {
	elements []rune
}

func (s *myStack) push(c rune) {
	s.elements = append(s.elements, c)
}
func (s *myStack) isEmpty() bool {
	return len(s.elements) == 0
}
func (s *myStack) peek() rune {
	return s.elements[len(s.elements)-1]
}
func (s *myStack) pop() rune {
	top := s.peek()
	s.elements = s.elements[:len(s.elements)-1]
	return top
}
